semantic-core/src/Data/Term.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE DeriveTraversable, FlexibleInstances, LambdaCase, MultiParamTypeClasses, QuantifiedConstraints, RankNTypes, ScopedTypeVariables, StandaloneDeriving, TypeOperators, UndecidableInstances #-}
+module Data.Term
+( Term(..)
+, Syntax(..)
+, iter
+, cata
+, interpret
+) where
+
+import Control.Effect.Carrier
+import Control.Monad (ap)
+import Control.Monad.Module
+import Data.Coerce (coerce)
+import Data.Functor.Const
+import Data.Scope
+
+data Term sig a
+  = Var a
+  | Term (sig (Term sig) a)
+
+deriving instance ( Eq a
+                  , RightModule sig
+                  , forall g x . (Eq  x, Monad g, forall y . Eq  y => Eq  (g y)) => Eq  (sig g x)
+                  )
+               => Eq  (Term sig a)
+deriving instance ( Ord a
+                  , RightModule sig
+                  , forall g x . (Eq  x, Monad g, forall y . Eq  y => Eq  (g y)) => Eq  (sig g x)
+                  , forall g x . (Ord x, Monad g, forall y . Eq  y => Eq  (g y)
+                                                , forall y . Ord y => Ord (g y)) => Ord (sig g x)
+                  )
+               => Ord (Term sig a)
+deriving instance (Show a, forall g x . (Show x, forall y . Show y => Show (g y)) => Show (sig g x)) => Show (Term sig a)
+
+deriving instance ( forall g . Foldable    g => Foldable    (sig g)) => Foldable    (Term sig)
+deriving instance ( forall g . Functor     g => Functor     (sig g)) => Functor     (Term sig)
+deriving instance ( forall g . Foldable    g => Foldable    (sig g)
+                  , forall g . Functor     g => Functor     (sig g)
+                  , forall g . Traversable g => Traversable (sig g)) => Traversable (Term sig)
+
+instance RightModule sig => Applicative (Term sig) where
+  pure = Var
+  (<*>) = ap
+
+instance RightModule sig => Monad (Term sig) where
+  Var  a >>= f = f a
+  Term t >>= f = Term (t >>=* f)
+
+instance RightModule sig => Carrier sig (Term sig) where
+  eff = Term
+
+
+class (HFunctor sig, forall g . Functor g => Functor (sig g)) => Syntax sig where
```

Comment:
This class is necessary if we want to fold over syntax, and unfortunately we currently still do kind of want to fold over syntax. I’m hoping to alleviate that, however.